Allow assign operator overload with old frontend

asodja opened the pull request.

asodja committed:
Add tests for old frontend

asodja committed:
Resolve assign in old frontend

asodja committed:
Resolve assign call on backend for old frontend

asodja committed:
Add some more tests for box testing

asodja committed:
Do not allow assign overload when there is delegate

asodja committed:
Fix IR backend to work correctly with class fields

demiurg906 commented on compiler/frontend/src/org/jetbrains/kotlin/types/expressions/ExpressionTypingVisitorForStatements.java:
```diff
@@ -441,6 +443,36 @@ protected KotlinTypeInfo visitAssignment(KtBinaryExpression expression, Expressi
+            temporaryForAssignmentOperation = TemporaryTraceAndCache.create(
+                    context, "trace to check assignment operation like '=' for", expression);
+            assignmentOperationDescriptors =
+                    components.callResolver.resolveBinaryCall(
```
If there is no `assign` then rhs will be resolved twice: firstly in `temporaryForAssignmentOperation` trace, and then in regular one at lin 479. This is bad, because it will lead to potentially exponential time of analysis in case if rhs contains some code block (lambda, if, when, try), which also contains such assignment

asodja replied:
Acknowledged. I will have to dig a bit, to better understand this resolution enough, to resolve this issue.

demiurg906 commented on compiler/frontend/src/org/jetbrains/kotlin/types/expressions/ExpressionTypingVisitorForStatements.java:
```diff
@@ -441,6 +443,36 @@ protected KotlinTypeInfo visitAssignment(KtBinaryExpression expression, Expressi
+            temporaryForAssignmentOperation.commit();
+            if (!KotlinTypeChecker.DEFAULT.equalTypes(components.builtIns.getUnitType(), assignmentOperationType)) {
+                KtSimpleNameExpression operationSign = expression.getOperationReference();
+                context.trace.report(ASSIGNMENT_OPERATOR_SHOULD_RETURN_UNIT.on(operationSign, assignmentOperationDescriptors.getResultingDescriptor(), operationSign));
```
Why do you report this diagnostic on call site instead of on declaration of the `assign` fun?

asodja replied:
For plusAssign&friends there is the same logic. Do you maybe know the reason why there is that kind of logic?

demiurg906 replied:
Oh, I get it. This is needed for a case when `plusAssign` came from java class, so we don't have a declaration site to report an error. So it's fine to leave it here

demiurg906 commented on compiler/frontend/src/org/jetbrains/kotlin/types/expressions/ExpressionTypingVisitorForStatements.java:
```diff
@@ -441,6 +443,36 @@ protected KotlinTypeInfo visitAssignment(KtBinaryExpression expression, Expressi
+                    context, "trace to check assignment operation like '=' for", expression);
+            assignmentOperationDescriptors =
+                    components.callResolver.resolveBinaryCall(
+                            context.replaceTraceAndCache(temporaryForAssignmentOperation).replaceScope(scope),
```
Do you check that `assign` fun really has `operator` modifier during call resolution?

Please add tests for this, like
```kotlin
operator fun String.assign(value: Int) {} // (1)

class Foo {
    fun String.assign(value: Int) {} // (2)

    fun test() {
        val s = "hello"
        s = 1 // should be resolved to (1)
    }
}

asodja replied:
Yeah, this is not checked and this test fails. It seems it's not that easy to do that for old frontend. I will have to dig into the code. Do you have any hint?

demiurg906 replied:
You can start from here
https://github.com/JetBrains/kotlin/blob/master/compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/components/ResolutionParts.kt#L775

asodja replied:
I believe this is now resolved, I also added a test.

demiurg906 commented on compiler/testData/codegen/box/operatorConventions/assignOperatorOverload.kt:
```diff
@@ -46,9 +57,9 @@ fun box(): String {
-    val nullCheck: NullCheck? = null
-    nullCheck?.x = "Fail"
-    if (nullCheckResult != "OK") return "Fail: $nullCheckResult"
+//    val nullCheck: NullCheck? = null
```
Why this code is commented out?

asodja replied:
This case did not work before. I added a test for unsafe diagnostics to frontend tests. But I have to double check this test if now it's working as intended.

asodja replied:
It seems to work ok now (the same as `plusAssign`). So if I have:
```
data class NullCheck(val x: NullCheckContainer)
data class NullCheckContainer(var value: String)

operator fun NullCheckContainer?.assign(value: String) {
    result = value
}
operator fun NullCheckContainer?.plusAssign(value: String) {
    result = value
}
```
and I do:
```
val nullCheck: NullCheck? = null
nullCheck?.x = "Fail"
nullCheck?.x += "Fail"
```

Then nothing is called in both cases.

asodja committed:
Move assign operator code behind a feature flag

asodja committed:
Select only methods with operator keyword for assign operator overload

asodja committed:
Make diagnostic correct when assign operator feature is not enabled

asodja committed:
Add test for diagnostics in case assign operator doesn't have Unit return type

asodja committed:
Don't use assign as convention operator and add additional tests

asodja commented:
@demiurg906 I put the feature behind the feature flag. I also added some more tests and resolved issue with incorrect resolved method when `operator` keyword is not used.  I also added tests that `val` still works correctly if feature is not enabled.

I believe method resolution now works correctly. Just the diagnostics could be better, now we always report "val cannot be reassigned" if no correct method is found. But we could do something better, but it would need some more work.

asodja committed:
Add additional test for null for assign operator overload

demiurg906 commented on compiler/testData/codegen/box/operatorConventions/assignOperatorOverload.kt:
```diff
@@ -1,5 +1,6 @@
 
 // TARGET_BACKEND: JVM_IR
+// !LANGUAGE:+AssignOperatorOverloadForJvmOldFrontend
```
Please also add test with disabled feature to ensure that new behaviour definitely not works if feature is not enabled

asodja replied:
There is this: [assignOperatorOverloadDisabled.kt](https://github.com/gradle/kotlin/pull/2/files/d74decff7b7a3bda859abd17f7db91fe2c3e522b..a64fca8f58e3166fed6e1d3ddda9bde9b82e06ab#diff-20b915b704de1013df4264283f347d7a1c8e3905a5d8f10f5a6f6c8468e272d5)

Should I do that somehow differently or maybe add more tests? 
Maybe also the name is not the best, maybe assignOperatorOverloadNotEnabled would be better, WDYT?

asodja replied:
Or you meant to add additional tests for box testing?

demiurg906 replied:
Sorry, I didn't notice that test

Naming is fine. I usually name such tests with `_enabled` and `_disabled` suffixes, but it's a question of taste

demiurg906 commented on compiler/cli/cli-common/src/org/jetbrains/kotlin/cli/common/arguments/CommonCompilerArguments.kt:
```diff
@@ -415,6 +415,9 @@ abstract class CommonCompilerArguments : CommonToolArguments() {
     @Argument(value = "-Xrender-internal-diagnostic-names", description = "Render internal names of warnings and errors")
     var renderInternalDiagnosticNames: Boolean by FreezableVar(false)
 
+    @Argument(value = "-Xenable-assign-operator-for-f10", description = "Enable assign operator overload for frontend 1.0")
```
Just `-Xassign-operator`

demiurg906 replied:
It's also required to put all your changes under a feature for FIR too

asodja replied:
I planned to have different flag for old and new frontend. That is because for old frontend, I have to have special handling in the backend. But maybe I can just check that if it's K2 compiler, I don't enable backend handling.

What are your thoughts?

demiurg906 replied:
Do you mean handling in psi2ir?
This component runs only for FE 1.0

asodja replied:
I meant psi2ir yes. Aaa, I didn't know that, I thought that that code always run if IR backend is enabled. Thx, that simplifes it, I will change it to `-Xassign-operator`.

demiurg906 replied:
FYI: for FIR we have similar part which is called fir2ir

asodja committed:
Rename enable assign flag to -Xassign-operator
